web/pdf_page_view.js

```diff
@@ -1006,20 +980,27 @@ class PDFPageView {
 
     // Keep the canvas hidden until the first draw callback, or until drawing
     // is complete when `!this.renderingQueue`, to prevent black flickering.
-    canvas.hidden = true;
     const hasHCM = !!(pageColors?.background && pageColors?.foreground);
+    const prevCanvas = this.canvas;
+    if (!prevCanvas && !hasHCM) {
+      canvasWrapper.append(canvas);
+    }
+    this.canvas = canvas;
 
     let showCanvas = isLastShow => {
       // In HCM, a final filter is applied on the canvas which means that
       // before it's applied we've normal colors. Consequently, to avoid to have
       // a final flash we just display it once all the drawing is done.
-      if (!hasHCM || isLastShow) {
-        canvas.hidden = false;
+      if ((!prevCanvas && !hasHCM) || isLastShow) {
+        if (prevCanvas) {
+          prevCanvas.replaceWith(canvas);
+          prevCanvas.width = prevCanvas.height = 0;
+        } else if (hasHCM) {
+          canvasWrapper.append(canvas);
+        }
         showCanvas = null; // Only invoke the function once.
```

Comment:
I'm a bit confused by the logic here. If `prevCanvas` exists and `hasHCM` is false, the first `(!prevCanvas && !hasHCM)` is `false` and thus we'll not show the new canvas until when `isLastShow` is true. This seems to contradict the comment above "Keep the canvas hidden until the first draw callback,"